optaplanner-core/src/test/java/org/optaplanner/core/impl/score/director/easy/EasyScoreDirectorTest.java

```diff
@@ -45,4 +56,33 @@ public void constraintMatchTotalsUnsupported() {
         return factory;
     }
 
+    @Test(expected = IllegalStateException.class)
+    public void shadowVariableCorruption() {
+        EasyScoreDirectorFactory<TestdataCorruptedShadowedSolution> scoreDirectorFactory =
+                new EasyScoreDirectorFactory<>((EasyScoreCalculator<TestdataCorruptedShadowedSolution>) (solution_) -> SimpleScore.valueOf(0));
+        scoreDirectorFactory.setSolutionDescriptor(TestdataCorruptedShadowedSolution.buildSolutionDescriptor());
+        scoreDirectorFactory.setInitializingScoreTrend(
+                InitializingScoreTrend.buildUniformTrend(InitializingScoreTrendLevel.ONLY_DOWN, 1));
+        EasyScoreDirector<TestdataCorruptedShadowedSolution> scoreDirector = scoreDirectorFactory.buildScoreDirector(false, false);
+
+        TestdataCorruptedShadowedSolution solution = new TestdataCorruptedShadowedSolution("s1");
+        TestdataValue v1 = new TestdataValue("v1");
+        TestdataValue v2 = new TestdataValue("v2");
+        solution.setValueList(Arrays.asList(v1, v2));
+        TestdataCorruptedShadowedEntity e1 = new TestdataCorruptedShadowedEntity("e1");
+        TestdataCorruptedShadowedEntity e2 = new TestdataCorruptedShadowedEntity("e2");
+        solution.setEntityList(Arrays.asList(e1, e2));
+        scoreDirector.setWorkingSolution(solution);
+
+        scoreDirector.assertShadowVariablesAreNotStale(SimpleScore.valueOfUninitialized(-2, 0), "NoChange");
```

Comment:
This line should never throw an exception, right? While the last line of this test should. If the test uses `@Test(expected = IllegalStateException.class)`, it will pass regardless where the exception has been thrown. I would suggest using try-catch with assertion of the `IllegalStateException` object for the last line of this test.

Reply:
Ok, I'll make the change, but I am not a fan of how the code will look like.
We really need to upgrade to JUnit 5, which has nice support for that.

Reply:
I don't like this code either, but until we migrate to jUnit5 or AssertJ, this is the only safe way to do such a checking.